postgresql/resource_postgresql_grant.go

```diff
@@ -296,6 +319,109 @@ WHERE grantee = $2
 	return nil
 }
 
+func readColumnRolePrivileges(txn *sql.Tx, d *schema.ResourceData) error {
+	objects := d.Get("objects").(*schema.Set)
+
+	missingColumns := d.Get("columns").(*schema.Set) // Getting columns from state.
+	// If the query returns a column, it is a removed from the missingColumns.
+
+	roleOID, err := getRoleOID(txn, d.Get("role").(string))
+	if err != nil {
+		return err
+	}
+
+	var rows *sql.Rows
+
+	// The following query is made up of 3 parts
+	// The first one simply aggregates all privileges on one column in one table into one line.
+	// The second part fetches all permissions on all columns for a given user & a given table in a give schema.
+	// The third part fetches all table-level permissions for the aforementioned table.
+	// Subtracting the third part from the second part allows us
+	// to get column-level privileges without those created by table-level privileges.
+	query := `
+SELECT table_name, column_name, array_agg(privilege_type) AS column_privileges
+FROM (
+  SELECT table_name, column_name, privilege_type
+  FROM information_schema.column_privileges
+    WHERE
+      grantee = $1
+    AND
+      table_schema = $2
+    AND
+      table_name = $3
+    AND
+      privilege_type = $6
+  EXCEPT
+  SELECT pg_class.relname, pg_attribute.attname, privilege_type AS table_grant
+  FROM pg_class
+  JOIN pg_namespace ON pg_namespace.oid = pg_class.relnamespace
+  LEFT JOIN (
+    SELECT acls.*
+    FROM
+      (SELECT relname, relnamespace, relkind, (aclexplode(relacl)).* FROM pg_class c) as acls
+    WHERE grantee=$4
+    ) privs
+  USING (relname, relnamespace, relkind)
+  LEFT JOIN pg_attribute ON pg_class.oid = pg_attribute.attrelid
+  WHERE nspname = $2 AND relkind = $5
+  )
+AS col_privs_without_table_privs
+GROUP BY col_privs_without_table_privs.table_name, col_privs_without_table_privs.column_name, col_privs_without_table_privs.privilege_type
+ORDER BY col_privs_without_table_privs.column_name
```

Comment:
@kda-jt  I got curious, so I dug into this query a little bit. I think the visibility issues that were being discussed in this PR could be avoided by only querying the pg catalog tables. I'm not sure if there's a particular reason you chose to source the privileges from `information_schema.column_privileges` and take a subtractive approach, so please correct me if I'm missing something here.

I've observed that column privileges are always visible in `pg_attribute.attacl`, even if the querying user does not have read access to the table. I'm unable to find explicit confirmation of this observation in the postgres docs, but I've tested it on several postgres databases (in RDS, 11 and 13, both aurora and standard RDS) and it appears the behavior is consistent.

Additionally, `pg_attribute.attacl` does not include entries for full-table grants, so using it as your source of truth also sidesteps the need to subtract table privilege grants.

With all that said, here's a query that I think faithfully produces the same output as your original, but only uses pg catalog tables:
```
SELECT relname AS table_name, attname AS column_name, array_agg(privilege_type) AS column_privileges
FROM (SELECT relname, attname, (aclexplode(attacl)).*
      FROM pg_class
               JOIN pg_namespace ON pg_class.relnamespace = pg_namespace.oid
               JOIN pg_attribute ON pg_class.oid = attrelid
      WHERE nspname = $2
        AND relname = $3
        AND relkind = $5)
         AS col_privs
         JOIN pg_roles ON pg_roles.oid = col_privs.grantee
WHERE rolname = '$1'
  AND privilege_type = '$6'
GROUP BY col_privs.relname, col_privs.attname, col_privs.privilege_type
ORDER BY col_privs.attname
```

If you like, I can send this as PR to your fork to update this PR.

It's worth an extra note that, RDS aside, this will be important for anyone who wants to perform management of their database roles using a non-superuser account.